### ISSUE TITLE
[popup windows] popup attached to text property stays visible even when text is hidden

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7736]
**Describe the bug**

A popup window attached to a text property might remain visible even when the text to which it's attached gets hidden, when maximizing another window and `'winminheight'` is set to 0.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        set lines=20
        repeat([''], 5)->setline(1)
        var buf = bufnr('%')
        prop_type_add('textprop', {bufnr: buf})
        prop_add(4, 1, {type: 'textprop', length: 25, bufnr: buf})
        popup_create('attached to text property', {textprop: 'textprop', highlight: 'ErrorMsg'})
        norm! G
        set sb wmh=0 nowrap
        h
        wincmd _
    EOF
    )

The popup window is visible in the help window.

**Expected behavior**

The popup window is not visible in the help window, because it's attached to a text which is hidden and displayed in another (minimized) window.

**Screenshots**

![gif](https://user-images.githubusercontent.com/8505073/105589227-0545c900-5d92-11eb-88e6-74b4706263ed.gif)

**Environment**

 - Vim version: 8.2 Included patches: 1-2396
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(363)

**Additional context**

The issue is influenced by the terminal geometry; in particular, its height.  Depending on its value, the issue might be reproducible or not.

---

For some reason, if `'wrap'` is off, the issue persists, but the position of the popup changes.  Its top line moves from the fourth line in the help window, to the first line:
```vim
vim9script
set lines=20
repeat([''], 5)->setline(1)
var buf = bufnr('%')
prop_type_add('textprop', {bufnr: buf})
prop_add(4, 1, {type: 'textprop', length: 25, bufnr: buf})
popup_create('attached to text property', {textprop: 'textprop', highlight: 'ErrorMsg'})
norm! G
set sb wmh=0
h
wincmd _
```
---

I found this issue while working on emulating virtual text.  I need popups to be attached to text properties.  But I also want `'winminheight'` to be set to 0, because I always have many splits and when I maximize the current window, I don't want to see 1 irrelevant line in every single split.

---

Workaround:  Maximize the window in 2 steps; that is:

   - temporarily reset `'winminheight'` to 1
   - maximize the window a first time
   - restore `'winminheight'` to its original 0 value
   - maximize the window a second time to finish the job

```vim
vim9script
set lines=20
repeat([''], 5)->setline(1)
var buf = bufnr('%')
prop_type_add('textprop', {bufnr: buf})
prop_add(4, 1, {type: 'textprop', length: 25, bufnr: buf})
popup_create('attached to text property', {textprop: 'textprop', highlight: 'ErrorMsg'})
norm! G
set sb wmh=0 nowrap
h
set wmh=1
wincmd _
set wmh=0
wincmd _
```
This is not a practical solution interactively, but it is a simple solution in a script; since I maximize the current window via autocmds, it's good enough for me.

**Comment by lacygoill** [comment]
> For some reason, if 'wrap' is off, the issue persists, but the position of the popup changes.

Sorry, I meant: "if 'wrap' is on".

**Comment by lacygoill** [comment]
It's clearly a bug, because if we reset `'winminheight'` to 1 instead of 0, the issue disappears:
```vim
vim9
set lines=20
repeat([''], 5)->setline(1)
var buf = bufnr('%')
prop_type_add('textprop', {bufnr: buf})
prop_add(4, 1, {type: 'textprop', length: 25, bufnr: buf})
popup_create('attached to text property', {textprop: 'textprop', highlight: 'ErrorMsg'})
norm! G
set sb wmh=1 nowrap
h
wincmd _
```
![gif](https://user-images.githubusercontent.com/8505073/105597619-4e971800-5d94-11eb-987a-0945db11bb7e.gif)

**Comment by lacygoill** [comment]
Fixed in patch [8.2.3193](https://github.com/vim/vim/releases/tag/v8.2.3193)

Since the latter included a test, I guess it's not necessary to turn the MWE in the OP into one more test.

---

This todo item is no longer necessary:

   > - Popup attached to text property stays visible when text is no longer
   >   visible. (#7736)
```diff
diff --git a/runtime/doc/todo.txt b/runtime/doc/todo.txt
index 80ea2a33e..0c6c64e13 100644
--- a/runtime/doc/todo.txt
+++ b/runtime/doc/todo.txt
@@ -131,8 +131,6 @@ Text properties:
   where property fits in.
   Or Should we let the textprop highlight overrule other (e.g. diff) highlight
   if the priority is above a certain value?  (#7392)
-- Popup attached to text property stays visible when text is no longer
-  visible. (#7736)
 - Popup attached to text property stays visible when text is deleted with
   "cc". (#7737)  "C" works OK.  "dd" also files in a buffer with a single
   line.
```

**Comment by brammool** [comment]
I'll remove the entry from the todo list, thanks.
